### PR TITLE
Adding version details to sidecar

### PR DESCRIPTION
**Pull request by saranyareddy24**
Fixes https://github.com/csi-addons/kubernetes-csi-addons/issues/377
Fixes https://github.com/csi-addons/kubernetes-csi-addons/issues/378

The output will be printed as:
Version: v0.6.0
Git Commit: 43ebc4baac5f8893a7d061bab9db78f66640b6b8
Go Version: go1.19.9
Compiler: gc
Platform: linux/amd64

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -56,6 +56,28 @@ ifneq ($(TAG),latest)
...
 endif
 
+# To get the GIT commit id
+ifndef GIT_COMMIT
+GIT_COMMIT=$(shell git rev-list -1 HEAD)
+endif
+
+# Fetch the git tag
+ifndef GIT_TAG
+GIT_TAG=$(shell git describe --tags --exact-match --abbrev=0 2>/dev/null)
```
Comment (nixpanic) [inline Makefile:66]: Makefile syntax makes it possible to drop the `ifndef`/`endif` by using the `?=` operator:

```make
GIT_TAG ?= $(shell git describe --tags --exact-match --abbrev=0 2>/dev/null)
```

File: Makefile [hunk 2]
```diff
@@ -56,6 +56,28 @@ ifneq ($(TAG),latest)
...
+endif
+
+# In the above command, if it fails to fetch the git tag, then this
+# command will fetch the short commit hash
+ifeq ($(GIT_TAG),)
+GIT_TAG=$(shell git rev-parse --short HEAD)
```
Comment (nixpanic) [inline Makefile:72]: Consider using `git describe` to get a more user-friendly tag-based line:
```
$ git describe --tags HEAD
v0.6.0-56-g342d4570
```

File: Makefile [hunk 3]
```diff
@@ -159,7 +181,7 @@ bundle-validate: container-cmd operator-sdk
 .PHONY: build
 build: generate fmt vet ## Build manager binary.
 	go build -o bin/manager cmd/manager/main.go
-	go build -o bin/sidecar sidecar/main.go
+	go build -ldflags '$(LDFLAGS)' -o bin/sidecar sidecar/main.go
```
Comment (nixpanic) [inline Makefile:184]: can you add this to the manager as well?

File: sidecar/internal/version/version.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2021 The Kubernetes-CSI-Addons Authors.
```
Comment (nixpanic) [inline sidecar/internal/version/version.go:2]: please correct the year

File: Makefile [hunk 5]
```diff
@@ -56,6 +56,28 @@ ifneq ($(TAG),latest)
...
+# In the above command, if it fails to fetch the git tag, then this
+# command will fetch the short commit hash
+ifeq ($(GIT_TAG),)
+GIT_TAG=$(shell git rev-parse --short HEAD)
+endif
+
+GO_PROJECT=github.com/csi-addons/kubernetes-csi-addons
+
+LDFLAGS ?=
+LDFLAGS += -X $(GO_PROJECT)/sidecar/internal/version.GitCommit=$(GIT_COMMIT)
```
Comment (nixpanic) [inline Makefile:78]: can you place this file in `internal/version/...` ? That way it can easily be used by both the manager and side-car.

**Comment by nixpanic** [comment]
Many thanks for you contribution! I think it will be very helpful :+1:

File: Makefile (hunk 2 above)
Comment (saranyareddy24) [reply Makefile:72]: I tried git describe command, but it fails if I am on main branch,
```
[ kubernetes-csi-addons]# git describe --tags HEAD
fatal: No names found, cannot describe anything.
[ kubernetes-csi-addons]# git branch
* main
```

File: Makefile (hunk 5 above)
Comment (saranyareddy24) [reply Makefile]: added for manager as well.
```
bin/manager -version
Version: v0.6.0
Git Commit: c433e0dac9a8e694e61c363f2e8aff5d3643b3a8
Go Version: go1.19.9
Compiler: gc
Platform: linux/amd64
```

File: sidecar/internal/version/version.go (hunk 4 above)
Comment (saranyareddy24) [reply sidecar/internal/version/version.go]: corrected to 2023.

File: Makefile (hunk 3 above)
Comment (saranyareddy24) [reply Makefile:182]: added to manager as well.

File: Makefile (hunk 1 above)
Comment (saranyareddy24) [reply Makefile]: replaced ifndef block with ?=

File: Makefile (hunk 2 above)
Comment (Madhu-1) [reply Makefile:68]: can you try on  origin main not on your local main `git checkout origin/main`?

Comment (saranyareddy24) [reply Makefile:68]: It worked. Thanks.

File: Makefile [hunk 6]
```diff
@@ -56,6 +56,19 @@ ifneq ($(TAG),latest)
 BUNDLE_VERSION ?= --version=$(shell sed s/^v// <<< $(TAG))
 endif
 
+# To get the GIT commit id
+GIT_COMMIT ?= $(shell git rev-list -1 HEAD)
+
+# Fetch the git tag
+GIT_TAG ?= $(shell git describe --tags HEAD=0 2>/dev/null)
```
Comment (nixpanic) [inline Makefile:63]: Thanks for checking with `git describe`. I get an error with `HEAD=0`, but `HEAD` works for me. If `git describe` is troublesome, using `git rev-list` as an alternative is ok too. Maybe this works?

```
GIT_TAG ?= $(shell git describe --tags HEAD 2>/dev/null || echo $(GIT_COMMIT))
```

Comment (saranyareddy24) [reply Makefile:63]: looks good. 
Although the version looks pretty big, if it's okay I can use short commit instead of using GIT_COMMIT

Comment (nixpanic) [reply Makefile:63]: great, thanks!

Comment (saranyareddy24) [reply Makefile]: updated

**Review by nixpanic — DISMISSED** [review]

**Comment by saranyareddy24** [comment]
@Rakshith-R & @yati1998 can you please check and approve

File: sidecar/main.go [hunk 7]
```diff
@@ -18,13 +18,13 @@ package main
...
 	"flag"
-	"time"
-
 	"github.com/csi-addons/kubernetes-csi-addons/internal/sidecar/service"
+	"github.com/csi-addons/kubernetes-csi-addons/internal/version"
 	"github.com/csi-addons/kubernetes-csi-addons/sidecar/internal/client"
 	"github.com/csi-addons/kubernetes-csi-addons/sidecar/internal/csiaddonsnode"
 	"github.com/csi-addons/kubernetes-csi-addons/sidecar/internal/server"
 	"github.com/csi-addons/kubernetes-csi-addons/sidecar/internal/util"
+	"time"
```
Comment (Rakshith-R) [inline sidecar/main.go:28]: Can you please move this back above?

File: sidecar/main.go [hunk 8]
```diff
@@ -47,12 +47,21 @@ func main() {
 		podNamespace = flag.String("namespace", "", "namespace of the Pod that contains this sidecar")
 		podUID       = flag.String("pod-uid", "", "UID of the Pod that contains this sidecar")
```
Comment (Rakshith-R) [inline sidecar/main.go:48]:
```suggestion
		podUID       = flag.String("pod-uid", "", "UID of the Pod that contains this sidecar")
		showVersion = flag.Bool("version", false, "Print Version details")
```

This is a bit more cleaner.

**Review by Rakshith-R — CHANGES_REQUESTED** [review]
Just a few small nits

File: sidecar/main.go (hunk 8 above)
Comment (saranyareddy24) [reply sidecar/main.go:48]: Thanks, modified accordingly.

File: sidecar/main.go (hunk 7 above)
Comment (saranyareddy24) [reply sidecar/main.go]: done.

**Review by Rakshith-R — APPROVED** [review]

**Comment by saranyareddy24** [comment]
can someone help me understand why build is failing.

**Comment by Rakshith-R** [comment]
> can someone help me understand why build is failing.

I think there is spacing issue in the import "time" pkg line. 
Can you try running `make test` locally?
or copy the spacing from the above "flag" pkg line and paste it there.

https://github.com/csi-addons/kubernetes-csi-addons/actions/runs/5385350011/jobs/9774541855?pr=385

and after amending your commit, check if that change still exists with `git show`

File: internal/version/version.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"fmt"
+	"runtime"
+)
+
+var (
+	// GitCommit tell the latest git commit image is built from.
+	GitCommit string
+	// SidecarVersion which will be sidecar release version.
```
Comment (Madhu-1) [inline internal/version/version.go:27]: The comment needs to be updated, currently it's referring to only sidecar.

**Review by Madhu-1 — COMMENTED** [review]
small nit, lgtm

File: sidecar/main.go [hunk 10]
```diff
@@ -18,9 +18,10 @@ package main
 
 import (
 	"flag"
-	"time"
+        "time"
```
Comment (Madhu-1) [inline sidecar/main.go:21]: you might have used space instead of tab here, please check

Comment (saranyareddy24) [reply sidecar/main.go]: yes, I changed to tab now.

File: internal/version/version.go (hunk 9 above)
Comment (saranyareddy24) [reply internal/version/version.go]: done.

**Review by Madhu-1 — APPROVED** [review]

**Review by Rakshith-R — APPROVED** [review]

**Comment by Madhu-1** [comment]
@mergifyio requeue

**Comment by Rakshith-R** [comment]
@mergifyio requeue

**Comment by Rakshith-R** [comment]
@saranyareddy24 

Can you please try logging into mergifyio 
or rebase this pr on top of latest main branch?
```
This pull request cannot be embarked for merge
The merge queue pull request can't be updated
Details:

Unable to update: user saranyareddy24 is unknown.

Please make sure saranyareddy24 has logged in Mergify dashboard.

```
[View more details on Mergify](https://dashboard.mergify.com/github/csi-addons/repo/kubernetes-csi-addons/queues?branch=main&queues=default&pull=385)

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Pull request merged**